### DISCUSSION TITLE
How to enable MAP_READ on a buffer produces by the `AsBindGroup` macro?

### DISCUSSION THREAD

**Discussion by davids91** [bevyengine/bevy#15159] [Q&A] [answered]
e.g. for the following type: 
```
#[derive(Resource, Clone, AsBindGroup, TypePath, ExtractResource)]
#[type_path = "..."]
pub struct MyData {
    #[storage(5, visibility(compute))]
    pub(crate) some_array: Vec<u32>,
}
```
Basically if I try to read from it I get the following error: 
```
wgpu_core::device::global: Buffer::map_async error: Usage flags BufferUsages(COPY_DST | STORAGE) for buffer Id(16,1,vk) do not contain required usage flags BufferUsages(MAP_READ)    
```

**Comment by JMS55** [comment]
You can't.

**Reply by davids91** [reply]
I believe the relevant part (found [here](https://github.com/gfx-rs/wgpu/issues/3305)) is : 
https://gpuweb.github.io/gpuweb/#dom-gpudevice-createbuffer

> If descriptor.[usage](https://gpuweb.github.io/gpuweb/#dom-gpubufferdescriptor-usage) contains [MAP_READ](https://gpuweb.github.io/gpuweb/#dom-gpubufferusage-map_read):
> 
> descriptor.[usage](https://gpuweb.github.io/gpuweb/#dom-gpubufferdescriptor-usage) must contain no other flags except [COPY_DST](https://gpuweb.github.io/gpuweb/#dom-gpubufferusage-copy_dst).

But then the only way to read the contents of a buffer is to copy it to another buffer, with map_read enabled, if I understand correctly.

**Reply by JMS55** [reply]
Yes, basically. Without mappable primary buffers, you need to issue a copy to the buffer, then map the buffer for reading (iirc, I rarely ever touch buffer mapping).

But AsBindGroup doesn't expose any of that anyways.

**Reply by davids91** [reply]
Thanks for the info! How about CPY_DST ? is that possible?

**Reply by JMS55** [reply, marked as answer]
No it is not. You can see the available options here: https://docs.rs/bevy/latest/bevy/render/render_resource/trait.AsBindGroup.html#deriving

**Reply by davids91** [reply]
Thank you!
